### PR TITLE
Allow running `cargo add foo` when a `foo.workspace = true` already exists

### PR DESCRIPTION
**Pull request by Muscraft**
Tracking issue: #8415
RFC: rust-lang/rfcs#2906

PRs in this RFC:
- Base Implementation
  - #10497
  - #10517 
  - #10538 
  - #10548 
- New Fields 
  - #10563
  - #10565 
- Update inheritance source 
  - #10564
- Optimizations
  - #10568 
- Documentation changes
  - #10584

This is currently a draft

This PR focuses on adding initial support to allow running `cargo add foo` when a `foo.workspace = true` already exists.

Changes:
- Added `WorkspaceSource` to allow for `foo.workspace = true`
- Added `MaybeWorkspace` to allow returning a `WorkspaceSource` when trying to `query` for a dependency or trying to get a `SourceId` for that query.

Remaining implementation work for the RFC
- `cargo-add` support, see [epage's comment](https://github.com/rust-lang/cargo/issues/8415#issuecomment-1075544790)
- Optimizations as needed

r? @epage

### CODE REVIEW

File: src/cargo/ops/cargo_add/dependency.rs [hunk 1]
```diff
@@ -822,6 +921,47 @@ impl std::fmt::Display for GitSource {
...
+/// Dependency from a workspace
+#[derive(Debug, Hash, PartialEq, Eq, Clone)]
+#[non_exhaustive]
+pub struct WorkspaceSource {
+    /// if workspace = true or false
+    pub workspace: bool,
+    /// features for this dependency not specified by workspace
+    pub features: Option<Vec<String>>,
+    /// if dependency is optional
+    pub optional: Option<bool>,
```
Comment (epage) [inline src/cargo/ops/cargo_add/dependency.rs:933]: This is unrelated to the source

File: src/cargo/ops/cargo_add/dependency.rs [hunk 2]
```diff
@@ -822,6 +921,47 @@ impl std::fmt::Display for GitSource {
...
+    /// Specify optional
+    pub fn set_optional(mut self, optional: impl Into<bool>) -> Self {
+        self.optional = Some(optional.into());
+        self
+    }
+}
+
+impl Display for WorkspaceSource {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        self.workspace.fmt(f)
```
Comment (epage) [inline src/cargo/ops/cargo_add/dependency.rs:961]: This will just print a bool.  We should probably just print `workspace`

**Comment by Muscraft** [comment]
I accidentally had my commits from an earlier branch that hadn't been merged under this one. It should be fixed now.

File: src/cargo/ops/cargo_add/dependency.rs [hunk 3]
```diff
@@ -822,6 +921,47 @@ impl std::fmt::Display for GitSource {
     }
 }
 
+/// Dependency from a workspace
+#[derive(Debug, Hash, PartialEq, Eq, Clone)]
+#[non_exhaustive]
+pub struct WorkspaceSource {
+    /// if workspace = true or false
+    pub workspace: bool,
```
Comment (epage) [inline src/cargo/ops/cargo_add/dependency.rs:929]: Can this ever be false?  Should we just have the presence of this assume `true`?

File: src/cargo/ops/cargo_add/dependency.rs [hunk 4]
```diff
@@ -480,6 +545,22 @@ impl Dependency {
...
+                        table.insert(
+                            "features",
+                            toml_edit::value(features.into_iter().collect::<toml_edit::Value>()),
+                        );
+                    }
+                    if let Some(optional) = src.optional {
+                        table.insert("optional", toml_edit::value(optional));
+                    }
+
+                    for key in ["path", "git", "branch", "tag", "rev", "version"] {
```
Comment (epage) [inline src/cargo/ops/cargo_add/dependency.rs:560]: We should delete all of the keys incompatible with `workspace = true`.

File: src/cargo/ops/cargo_add/dependency.rs [hunk 5]
```diff
@@ -480,6 +545,22 @@ impl Dependency {
                         table.remove(key);
                     }
                 }
+                Some(Source::Workspace(src)) => {
+                    table.insert("workspace", toml_edit::value(src.workspace));
```
Comment (epage) [inline src/cargo/ops/cargo_add/dependency.rs:549]: The other branches should remove `"workspace"`

File: src/cargo/ops/cargo_add/dependency.rs [hunk 6]
```diff
@@ -397,6 +447,21 @@ impl Dependency {
...
                     }
+                    Some(Source::Workspace(src)) => {
+                        table.set_dotted(true);
+                        table.insert("workspace", src.workspace.into());
+                        if let Some(features) = src.features.as_deref() {
+                            table.insert(
+                                "features",
+                                features.into_iter().collect::<toml_edit::Value>(),
+                            );
+                            table.set_dotted(false);
```
Comment (epage) [inline src/cargo/ops/cargo_add/dependency.rs:458]: Another option is to match for this case specially like we do for `foo = "version"`

File: src/cargo/ops/cargo_add/dependency.rs [hunk 7]
```diff
@@ -271,6 +292,35 @@ impl Dependency {
                         invalid_type(key, "version", version.type_name(), "string")
                     })?);
                     src.into()
+                } else if let Some(workspace) = table.get("workspace") {
+                    let mut src = WorkspaceSource::new(workspace.as_bool().ok_or_else(|| {
+                        invalid_type(key, "workspace", workspace.type_name(), "bool")
+                    })?);
+                    if !src.workspace {
+                        anyhow::bail!("workspace was false for `{key}`")
```
Comment (epage) [inline src/cargo/ops/cargo_add/dependency.rs:300]: Would it be more clear to say `"{key}.workspace = false is unsupported"`

File: src/cargo/ops/cargo_add/mod.rs [hunk 8]
```diff
@@ -389,31 +391,38 @@ fn get_latest_dependency(
     registry: &mut PackageRegistry<'_>,
 ) -> CargoResult<Dependency> {
     let query = dependency.query(config)?;
-    let possibilities = loop {
-        let fuzzy = true;
-        match registry.query_vec(&query, fuzzy) {
-            std::task::Poll::Ready(res) => {
-                break res?;
+    match query {
+        MaybeWorkspace::Workspace(_) => Err(anyhow!("query was workspace")),
```
Comment (epage) [inline src/cargo/ops/cargo_add/mod.rs:395]: I might go as far as an `unreachable!("registry depedencies required, found a workspace dependency")`

File: src/cargo/ops/cargo_add/mod.rs [hunk 9]
```diff
@@ -422,36 +431,41 @@ fn select_package(
...
-                .source()
-                .expect("source should be resolved before here");
-            anyhow::bail!("the crate `{dependency}` could not be found at `{source}`")
-        }
-        1 => {
-            let mut dep = Dependency::from(&possibilities[0]);
-            if let Some(reg_name) = dependency.registry.as_deref() {
-                dep = dep.set_registry(reg_name);
+    match query {
+        MaybeWorkspace::Workspace(_) => Err(anyhow!("query was workspace")),
```
Comment (epage) [inline src/cargo/ops/cargo_add/mod.rs:435]: ditto
```
unreachable!("path or git dependency expected, found workspace dependency");
```

File: src/cargo/ops/cargo_add/mod.rs [hunk 10]
```diff
@@ -518,28 +532,44 @@ fn populate_available_features(
...
-            std::task::Poll::Ready(res) => {
-                break res?;
+    match query {
+        MaybeWorkspace::Workspace(workspace) => {
+            if let Some(features) = workspace.features {
+                let mut b_tree: BTreeMap<String, Vec<String>> = BTreeMap::new();
+                for feature in features {
+                    b_tree.insert(feature, vec![]);
+                }
+                dependency = dependency.set_available_features(b_tree)
```
Comment (epage) [inline src/cargo/ops/cargo_add/mod.rs:542]: Don't we need to be looking up the workspace's dependency and getting its query?

File: tests/testsuite/cargo_add.rs [hunk 11]
```diff
@@ -1504,6 +1504,74 @@ fn overwrite_inline_features() {
...
+}
+
+#[cargo_test]
+fn overwrite_inherit_noop() {
+    init_registry();
+    let project_root = project_from_template("tests/snapshots/add/overwrite_inherit_noop.in");
+    let cwd = &project_root;
+
+    cargo_command()
+        .env("__CARGO_TEST_CHANNEL_OVERRIDE_DO_NOT_USE_THIS", "nightly")
```
Comment (epage) [inline tests/testsuite/cargo_add.rs:1537]: #10581 will add a `masquerade_as_nightly_cargo()` function to `Command`.

File: src/cargo/ops/cargo_add/dependency.rs (hunk 3 above)
Comment (Muscraft) [reply src/cargo/ops/cargo_add/dependency.rs:929]: That is up to you. I don't think it should be false ever since we check if it's false before we create it.

File: src/cargo/ops/cargo_add/mod.rs (hunk 10 above)
Comment (Muscraft) [reply src/cargo/ops/cargo_add/mod.rs:542]: Possibly, this was one area that I wasn't sure how we should be going about this. I know that this makes it so `foo = { workspace = true, features = ["test"]}` works.

File: src/cargo/ops/cargo_add/dependency.rs (hunk 1 above)
Comment (epage) [reply src/cargo/ops/cargo_add/dependency.rs:933]: See
```rust
pub struct Dependency {
    /// The name of the dependency (as it is set in its `Cargo.toml` and known to crates.io)
    pub name: String,
    /// Whether the dependency is opted-in with a feature flag
    pub optional: Option<bool>,

    /// List of features to add (or None to keep features unchanged).
    pub features: Option<IndexSet<String>>,
    /// Whether default features are enabled
    pub default_features: Option<bool>,

    /// Where the dependency comes from
    pub source: Option<Source>,
    /// Non-default registry
    pub registry: Option<String>,

    /// If the dependency is renamed, this is the new name for the dependency
    /// as a string.  None if it is not renamed.
    pub rename: Option<String>,

    /// Features that are exposed by the dependency
    pub available_features: BTreeMap<String, Vec<String>>,
}
```

File: src/cargo/ops/cargo_add/dependency.rs (hunk 3 above)
Comment (epage) [reply src/cargo/ops/cargo_add/dependency.rs:929]: Let's just have this be a unit struct for now.  We can always add fields later

File: tests/testsuite/cargo_add.rs (hunk 11 above)
Comment (Muscraft) [reply tests/testsuite/cargo_add.rs:1537]: I saw that and put this in until that PR is merged or we plan to merge this one and then have a cleanup PR that fixes it

File: src/cargo/ops/cargo_add/dependency.rs (hunk 1 above)
Comment (epage) [reply src/cargo/ops/cargo_add/dependency.rs:933]: Correct

File: src/cargo/ops/cargo_add/mod.rs (hunk 10 above)
Comment (epage) [reply src/cargo/ops/cargo_add/mod.rs:542]: We have two feature lists
- What the user specified
- What the crate defines

This code is putting the user-specified features into the list of what the crate defines so anything the user enters will always be valid.

File: src/cargo/ops/cargo_add/dependency.rs (hunk 6 above)
Comment (epage) [reply src/cargo/ops/cargo_add/dependency.rs:458]: See
```rust
        let table: toml_edit::Item = match (
            self.optional.unwrap_or(false),
            self.features.as_ref(),
            self.default_features.unwrap_or(true),
            self.source.as_ref(),
            self.registry.as_ref(),
            self.rename.as_ref(),
        ) {
            // Extra short when version flag only
            (
                false,
                None,
                true,
                Some(Source::Registry(RegistrySource { version: v })),
                None,
                None,
            ) => toml_edit::value(v),
            // Other cases are represented as an inline table
            (_, _, _, _, _, _) => {
```
(wish github would let me comment on code outside of the diffs)

**Comment by Muscraft** [comment]
Closing as #10606 includes this commit
